src/openvic-simulation/types/Colour.hpp

```diff
@@ -1,42 +1,218 @@
-
-	constexpr colour_t float_to_colour_byte(float f, float min = 0.0f, float max = 1.0f) {
-		return static_cast<colour_t>(std::clamp(min + f * (max - min), min, max) * 255.0f);
-	}
+	template<typename ValueT, typename ComponentReturnT, colour_traits Traits = colour_traits<ValueT, ComponentReturnT> {}>
+	struct basic_colour_t {
+		static constexpr auto traits = Traits;
+		using value_type = typename decltype(traits)::value_type;
+		using component_return_type = typename decltype(traits)::component_return_type;
 
-	constexpr colour_t fraction_to_colour_byte(int n, int d, float min = 0.0f, float max = 1.0f) {
-		return float_to_colour_byte(static_cast<float>(n) / static_cast<float>(d), min, max);
-	}
+		value_type value;
 
-	constexpr colour_t float_to_alpha_value(float a) {
-		return float_to_colour_byte(a) << 24;
-	}
+		/* When colour_t is used as an identifier, NULL_COLOUR is disallowed
+		 * and should be reserved as an error value.
+		 * When colour_t is used in a purely graphical context, NULL_COLOUR
+		 * should be allowed.
+		 */
+		static constexpr value_type null = traits.null;
+		static constexpr value_type component = traits.component;
+		static constexpr value_type max_rgb = traits.max_rgb;
+		static constexpr value_type max_argb = traits.max_argb;
 
-	constexpr float colour_byte_to_float(colour_t colour) {
-		return std::clamp(static_cast<float>(colour) / 255.0f, 0.0f, 1.0f);
-	}
+		static constexpr value_type blue_shift = traits.blue_shift;
+		static constexpr value_type green_shift = traits.green_shift;
+		static constexpr value_type red_shift = traits.red_shift;
+		static constexpr value_type alpha_shift = traits.alpha_shift;
 
-	inline std::string colour_to_hex_string(colour_t colour, bool alpha = false) {
-		std::ostringstream stream;
-		stream << std::uppercase << std::hex << std::setfill('0') << std::setw(!alpha ? 6 : 8) << colour;
-		return stream.str();
-	}
+		constexpr basic_colour_t() {}
+		constexpr basic_colour_t(value_type p_value) : value(p_value) {}
+		constexpr explicit operator value_type() const {
+			return value;
+		}
+
+		static constexpr basic_colour_t from_float(float f, float min = 0.0f, float max = 1.0f) {
+			return static_cast<value_type>(std::clamp(min + f * (max - min), min, max) * 255.0f);
+		}
+
+		static constexpr basic_colour_t from_fraction(int n, int d, float min = 0.0f, float max = 1.0f) {
+			return from_float(static_cast<float>(n) / static_cast<float>(d), min, max);
+		}
+
+		static constexpr basic_colour_t from_float_to_alpha(float a) {
+			return from_float(a) << 24;
+		}
+
+		inline std::string to_hex_string(bool alpha = false) const {
+			using namespace std::string_view_literals;
+			static constexpr std::string_view digits = "0123456789ABCDEF"sv;
+
+			const std::int8_t length = alpha ? 8 : 6;
+			std::array<char, 8> address;
+
+			for (std::size_t i = 0, j = (length - 1) * 4; i < length; ++i, j -= 4) {
+				address[i] = digits[(value >> j) & 0x0f];
+			}
+			return { address.data() };
```

Comment:
Very nice, love a good two-variable for loop. I assume `address` default zero-initialises its memory?

Reply:
std::array default initializes its memory, since char is a integral type, its default value is 0.